File: ios/App/App/Wearables/Garmin/GarminDeviceStore.swift
import Foundation

/*
 UNVERIFIED / NOT WIRED INTO THE BUILD. This file is not in project.pbxproj's
 Sources build phase — see garmin/README.md and the Garmin Phase 1 plan
 step 5. Add it once the Garmin Connect IQ Mobile SDK for iOS is linked.

 Persists which Garmin device Caddy should treat as the wearable target.
 Unlike Apple Watch — where WCSession always addresses "the paired Watch,"
 singular — a phone may have Garmin Connect paired to several devices, and
 Caddy must not silently send a round to all of them. This store holds the
 one deliberate choice, mirroring the field list the Garmin Phase 1 plan
 step 5 calls for.
*/
final class GarminDeviceStore {
    struct SelectedDevice: Equatable {
        let deviceId: String
        let deviceName: String
        let model: String
    }

    enum ConnectionState: String {
        case unknown, notConnected, connected, unavailable
    }

    private static let deviceIdKey = "GarminSelectedDeviceIdV1"
    private static let deviceNameKey = "GarminSelectedDeviceNameV1"
    private static let modelKey = "GarminSelectedDeviceModelV1"
    private static let connectionStateKey = "GarminLastKnownConnectionStateV1"

    private let defaults: UserDefaults

    init(defaults: UserDefaults = .standard) {
        self.defaults = defaults
    }

    var selectedDevice: SelectedDevice? {
        guard let id = defaults.string(forKey: Self.deviceIdKey),
              let name = defaults.string(forKey: Self.deviceNameKey) else { return nil }
        let model = defaults.string(forKey: Self.modelKey) ?? ""
        return SelectedDevice(deviceId: id, deviceName: name, model: model)
    }

    var lastKnownConnectionState: ConnectionState {
        ConnectionState(rawValue: defaults.string(forKey: Self.connectionStateKey) ?? "") ?? .unknown
    }

    /* Caddy should deliberately choose which Garmin is the Caddy wearable —
       this is never set automatically from device discovery. The caller
       (a future device-picker UI) supplies the choice explicitly. */
    func select(deviceId: String, deviceName: String, model: String) {
        defaults.set(deviceId, forKey: Self.deviceIdKey)
        defaults.set(deviceName, forKey: Self.deviceNameKey)
        defaults.set(model, forKey: Self.modelKey)
    }

    func clearSelection() {
        defaults.removeObject(forKey: Self.deviceIdKey)
        defaults.removeObject(forKey: Self.deviceNameKey)
        defaults.removeObject(forKey: Self.modelKey)
    }

    func recordConnectionState(_ state: ConnectionState) {
        defaults.set(state.rawValue, forKey: Self.connectionStateKey)
    }
}

File: android/app/src/main/java/com/claritygolf/caddy/wearables/garmin/GarminDeviceStore.java
package com.claritygolf.caddy.wearables.garmin;

import android.content.Context;
import android.content.SharedPreferences;

/**
 * Persists which Garmin device Caddy should treat as the wearable target.
 * Mirrors ios/App/App/Wearables/Garmin/GarminDeviceStore.swift field for
 * field, backed by SharedPreferences instead of UserDefaults (see the
 * Garmin-on-Android scope note this session wrote earlier: Android's
 * wearable target is Garmin only, no Wear OS, so there is exactly one
 * transport and this store has exactly one caller).
 *
 * Caddy chooses which Garmin is the Caddy wearable deliberately — this store
 * never auto-selects from device discovery. A future device-picker UI
 * supplies the choice explicitly via {@link #select}.
 */
public final class GarminDeviceStore {
    private static final String PREFS_NAME = "GarminDeviceStoreV1";
    private static final String KEY_DEVICE_ID = "deviceId";
    private static final String KEY_DEVICE_NAME = "deviceName";
    private static final String KEY_MODEL = "model";
    private static final String KEY_CONNECTION_STATE = "connectionState";

    public enum ConnectionState {
        UNKNOWN, NOT_CONNECTED, CONNECTED, UNAVAILABLE
    }

    public static final class SelectedDevice {
        public final String deviceId;
        public final String deviceName;
        public final String model;

        public SelectedDevice(String deviceId, String deviceName, String model) {
            this.deviceId = deviceId;
            this.deviceName = deviceName;
            this.model = model;
        }
    }

    private final SharedPreferences prefs;

    public GarminDeviceStore(Context context) {
        this.prefs = context.getApplicationContext().getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE);
    }

    public SelectedDevice getSelectedDevice() {
        String id = prefs.getString(KEY_DEVICE_ID, null);
        String name = prefs.getString(KEY_DEVICE_NAME, null);
        if (id == null || name == null) { return null; }
        String model = prefs.getString(KEY_MODEL, "");
        return new SelectedDevice(id, name, model);
    }

    public ConnectionState getLastKnownConnectionState() {
        String raw = prefs.getString(KEY_CONNECTION_STATE, ConnectionState.UNKNOWN.name());
        try {
            return ConnectionState.valueOf(raw);
        } catch (IllegalArgumentException e) {
            return ConnectionState.UNKNOWN;
        }
    }

    public void select(String deviceId, String deviceName, String model) {
        prefs.edit()
            .putString(KEY_DEVICE_ID, deviceId)
            .putString(KEY_DEVICE_NAME, deviceName)
            .putString(KEY_MODEL, model)
            .apply();
    }

    public void clearSelection() {
        prefs.edit()
            .remove(KEY_DEVICE_ID)
            .remove(KEY_DEVICE_NAME)
            .remove(KEY_MODEL)
            .apply();
    }

    public void recordConnectionState(ConnectionState state) {
        prefs.edit().putString(KEY_CONNECTION_STATE, state.name()).apply();
    }
}
